File: Linear_Data_Structures/Linked List/singly_linked_list.py
"""
Singly linked list implementation and demo
"""

class Node:
    """Node for singly linked list"""
    def __init__(self, data):
        self.data = data
        self.next = None


class SinglyLinkedList:
    """Singly linked list with basic operations."""
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def insert_at_head(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
        if self.tail is None:
            self.tail = new_node
        self.size += 1

    def insert_at_tail(self, data):
        new_node = Node(data)
        if self.tail:
            self.tail.next = new_node
            self.tail = new_node
        else:
            self.head = self.tail = new_node
        self.size += 1

    def search(self, key):
        current = self.head
        while current:
            if current.data == key:
                return True
            current = current.next
        return False

    def delete(self, key):
        prev = None
        current = self.head
        while current:
            if current.data == key:
                if prev:
                    prev.next = current.next
                else:
                    self.head = current.next
                if current == self.tail:
                    self.tail = prev
                self.size -= 1
                return True
            prev = current
            current = current.next
        return False

    def traverse(self):
        current = self.head
        while current:
            yield current.data
            current = current.next

    def __len__(self):
        return self.size


if __name__ == "__main__":
    lst = SinglyLinkedList()
    for v in [10, 20, 30]:
        lst.insert_at_tail(v)
    print("Initial list:", list(lst.traverse()))

    lst.insert_at_head(5)
    print("After insert_at_head(5):", list(lst.traverse()))

    lst.insert_at_tail(40)
    print("After insert_at_tail(40):", list(lst.traverse()))

    print("Search 30:", lst.search(30))
    print("Delete 20:", lst.delete(20))
    print("After delete(20):", list(lst.traverse()))
    print("Length:", len(lst))

File: Linear_Data_Structures/Linked List/singly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

/*
 * Singly linked list implementation in C
 * Simple demo: insert head/tail, search, delete, traverse
 */

typedef struct Node {
    int data;
    struct Node *next;
} Node;

Node* create_node(int data) {
    Node *n = (Node*)malloc(sizeof(Node));
    n->data = data;
    n->next = NULL;
    return n;
}

void insert_at_head(Node **head, Node **tail, int data) {
    Node *n = create_node(data);
    n->next = *head;
    *head = n;
    if (*tail == NULL) *tail = n;
}

void insert_at_tail(Node **head, Node **tail, int data) {
    Node *n = create_node(data);
    if (*tail) {
        (*tail)->next = n;
        *tail = n;
    } else {
        *head = *tail = n;
    }
}

int search(Node *head, int key) {
    Node *cur = head;
    while (cur) {
        if (cur->data == key) return 1;
        cur = cur->next;
    }
    return 0;
}

int delete_value(Node **head, Node **tail, int key) {
    Node *cur = *head;
    Node *prev = NULL;
    while (cur) {
        if (cur->data == key) {
            if (prev) prev->next = cur->next;
            else *head = cur->next;
            if (cur == *tail) *tail = prev;
            free(cur);
            return 1;
        }
        prev = cur;
        cur = cur->next;
    }
    return 0;
}

void traverse(Node *head) {
    Node *cur = head;
    printf("[");
    while (cur) {
        printf("%d", cur->data);
        if (cur->next) printf(", ");
        cur = cur->next;
    }
    printf("]\n");
}

void free_list(Node *head) {
    Node *cur = head;
    while (cur) {
        Node *next = cur->next;
        free(cur);
        cur = next;
    }
}

int main() {
    Node *head = NULL, *tail = NULL;
    insert_at_tail(&head, &tail, 10);
    insert_at_tail(&head, &tail, 20);
    insert_at_head(&head, &tail, 5);
    printf("List after inserts: ");
    traverse(head);

    printf("Search 20: %s\n", search(head, 20) ? "found" : "not found");
    delete_value(&head, &tail, 20);
    printf("After delete 20: ");
    traverse(head);

    free_list(head);
    return 0;
}
